### DISCUSSION TITLE
RFC `RuleTiles / AutoTiles`

### DISCUSSION THREAD

**Discussion by frederickjjoubert** [StarArawn/bevy_ecs_tilemap#326] [Ideas]
# RFC `RuleTiles / AutoTiles`

## Summary

Rule Tiles / Auto Tiles are tiles that automatically update their sprites based on their neighbors / adjacent tiles. Rules supplied by the developer govern which neighbors are allowed or not allowed, and based on their presence, the sprite is updated to match a particular rule.

## Motivation

This is a tilemap plugin, and the development of games with tilemaps require rule tiles for developer convenience and development speed. Rule tiles allow developers to build complex tilemaps quickly, both during development for static tilemaps, and at run-time for dynamic tilemaps.

Other engines such as [Unity](https://docs.unity3d.com/Packages/com.unity.2d.tilemap.extras@3.0/manual/RuleTile.html) and [Godot](https://docs.godotengine.org/en/stable/tutorials/2d/using_tilemaps.html#autotiles) already support rule tiles / auto tiles.

## User-facing explanation

I would imagine Rule Tiles being inserted as a component similar to the current [TileBundle](https://docs.rs/bevy_ecs_tilemap/latest/bevy_ecs_tilemap/tiles/struct.TileBundle.html) and [AnimatedTile](https://docs.rs/bevy_ecs_tilemap/latest/bevy_ecs_tilemap/tiles/struct.AnimatedTile.html)

Something like this following the current API and [basic example](https://github.com/StarArawn/bevy_ecs_tilemap/blob/main/examples/basic.rs):

```rust
for x in 0..tilemap_size.x {
	for y in 0..tilemap_size.y {
		let tile_pos = TilePos { x, y };
		let tile_entity = commands
			.spawn()
			.insert_bundle(RuleTileBundle {
				position: tile_pos,
				tilemap_id: TilemapId(tilemap_entity),
				default_sprite: ...
				rules: [
					...
				]
				..Default::default()
		})
		.id();
	tile_storage.set(&tile_pos, tile_entity);
	}
}
```

where `rules` (`Vec<(Sprite, Rule)>`) could have the following data structures:

```rust
pub struct RuleTile {  
    pub default_sprite: Sprite,  
    pub rules: Vec<(Rule, Sprite)>,  
}  
  
pub struct Rule {  
    pub nw_slot: Slot,  
    pub n_slot: Slot,  
    pub ne_slot: Slot,  
    pub w_slot: Slot,  
    pub e_slot: Slot,  
    pub sw_slot: Slot,  
    pub s_slot: Slot,  
    pub se_slow: Slot,  
}

pub enum Slot {  
    Empty,  
    Filled,  
    Any,  
}  
```

## Implementation strategy

TBD

Further discussion is needed.

## Drawbacks

I don't see any drawbacks to add this feature, it is a staple in other engines that feature tilemaps. The biggst consideration is development time and complexity.

## Rationale and alternatives

TBD

Further discussion is needed.

### Unity

[Unity Rule Tiles](https://docs.unity3d.com/Packages/com.unity.2d.tilemap.extras@3.0/manual/RuleTile.html)

Code: https://github.com/Unity-Technologies/2d-extras/blob/master/Runtime/Tiles/RuleTile/RuleTile.cs

Custom Rule Tiles: https://docs.unity3d.com/Packages/com.unity.2d.tilemap.extras@1.6/manual/CustomRulesForRuleTile.html

### Godot

[Godot AutoTiles](https://docs.godotengine.org/en/stable/tutorials/2d/using_tilemaps.html#autotiles)

### Game Maker Studio

https://gamemaker.io/en/blog/practical-guide-tiles
https://gamemaker.io/en/tutorials/tile-set-editor

### Other Articles

https://www.boristhebrave.com/2013/07/14/tileset-roundup/?q=tutorials/tileset-roundup
	- Different Tileset behaviors

## Unresolved questions

- I am not sure what the scope of this would be / time to develop.
- I am not sure how this would integrate with `bevy_ecs_tilemap` "under the hood" as far as rendering / draw performance goes, that is outside of my current knowledge.
- I am not sure how this would work with non-standard 2d grids such as `hex` or `iso`.

## Future possibilities

I understand this is a big feature to add to `bevy_ecs_tilemap`, but I do think it is a necessary long term feature to add, especially if `bevy_ecs_tilemap` ends up becoming part of `bevy` itself in the long term.

**Comment by frederickjjoubert** [comment]
Working Example of Autotiling on top of `bevy_ecs_tilemap` https://github.com/frederickjjoubert/bevy_ecs_tilemap_autotile_example

**Comment by doonv** [comment]
I would love it if this got implemented.

**Comment by hjk321** [comment]
Is this still under consideration?
